src/python/pants/option/parser.py

```diff
@@ -221,13 +265,64 @@ def add_flag_val(v):
 
       setattr(namespace, dest, val)
 
-    # See if there are any unconsumed flags remaining.
+    # See if there are any unconsumed flags remaining, and if so, raise a ParseError.
     if flag_value_map:
-      raise ParseError('Unrecognized command line flags on {}: {}'.format(
-        self._scope_str(), ', '.join(flag_value_map.keys())))
+      self._raise_error_for_invalid_flag_names(
+        flag_value_map,
+        all_scoped_flag_names=list(get_all_scoped_flag_names()),
+        levenshtein_max_distance=levenshtein_max_distance)
 
     return namespace
 
+  def _raise_error_for_invalid_flag_names(self, flag_value_map, all_scoped_flag_names,
+                                          levenshtein_max_distance):
+    """Identify similar option names to unconsumed flags and raise a ParseError with those names."""
+    matching_flags = {}
+    for flag_name in flag_value_map.keys():
+      # We will be matching option names without their leading hyphens, in order to capture both
+      # short and long-form options.
+      flag_minus_leading_dashes = re.sub(r'^-+', '', flag_name)
+      scoped_name = (
+        '{}-{}'.format(self.scope, flag_minus_leading_dashes)
+        if self.scope != GLOBAL_SCOPE
+        else flag_minus_leading_dashes)
+      suffix_matching_option_names = []
+      levenshtein_matching_option_names = defaultdict(list)
+      for other_name in all_scoped_flag_names:
+        other_name_no_dashes = re.sub(r'^-+', '', other_name)
+        # If the invalid option on the command line is the ending of another option, display
+        # that. This can occur when an option is used without the appropriate prefix scope.
+        if other_name_no_dashes.endswith(scoped_name):
+          suffix_matching_option_names.append(other_name)
+        else:
```

Comment:
Should this else clause be there? I think it's possible to have a case where the first few letters are spelled wrong, but the suffix is right _and_ the scope is missing.

```bash
./pants --interprete-search-paths
```

This I think would result in the `suffix_matching_option_names` clause triggering. It should also probably trigger the mispelling? Although I don't think Levenshtein would ever realistically be triggered in this case, because missing the prefix means the distance is almost certainly going to be greater than 2.

Either way, I'd love a test case for suffix missing _and_ typo in the prefix. I think a reasonable return would be for `--interprete-search-paths` to suggest `--python-setup-interpreter-search-paths`, which it seems we currently would do.

Reply:
The logic had to be redone, but this was fixed! The test case used is the invalid option `--simple-sam`.